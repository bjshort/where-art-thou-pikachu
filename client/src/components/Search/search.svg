<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="20" height="20" viewBox="0 0 20 20">
    <defs>
        <path id="a" d="M13.42 15.122a8.369 8.369 0 0 1-5.02 1.663c-4.639 0-8.4-3.758-8.4-8.393C0 3.757 3.761 0 8.4 0c4.64 0 8.402 3.757 8.402 8.392a8.35 8.35 0 0 1-1.682 5.038L20 18.305 18.303 20l-4.883-4.878zm-5.02-.735c3.315 0 6.001-2.684 6.001-5.995a5.998 5.998 0 0 0-6-5.994c-3.314 0-6 2.684-6 5.994a5.998 5.998 0 0 0 6 5.995z"/>
    </defs>
    <g fill="none" fill-rule="evenodd">
        <mask id="b" fill="#fff">
            <use xlink:href="#a"/>
        </mask>
        <use fill="#333" fill-rule="nonzero" xlink:href="#a"/>
    </g>
</svg>
