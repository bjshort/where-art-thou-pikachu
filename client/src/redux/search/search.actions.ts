export const UPDATE_QUERY = 'SEARCH/UPDATE_QUERY';
export const FETCH_POKEMON_REQUESTED = 'SEARCH/FETCH_POKEMON_REQUESTED';
export const FETCH_POKEMON_FAILED = 'SEARCH/FETCH_POKEMON_FAILED';
export const FETCH_POKEMON_SUCCESS = 'SEARCH/FETCH_POKEMON_SUCCESS';

export const TOGGLE_TRANSLATION = 'SEARCH/TOGGLE_TRANSLATION';

export function updateQuery(query: string) {
  return {
    type: UPDATE_QUERY,
    payload: query,
  } as const;
}

export function fetchPokemonRequested() {
  return {
    type: FETCH_POKEMON_REQUESTED,
  } as const;
}

export function fetchPokemonFailed(error: string) {
  return {
    type: FETCH_POKEMON_FAILED,
    payload: error,
  } as const;
}

export function fetchPokemonSuccess(id: string) {
  return {
    type: FETCH_POKEMON_SUCCESS,
    payload: id,
  } as const;
}

export function toggleTranslation() {
  return {
    type: TOGGLE_TRANSLATION,
  } as const;
}

export type SearchActions = ReturnType<
  | typeof updateQuery
  | typeof fetchPokemonRequested
  | typeof fetchPokemonFailed
  | typeof fetchPokemonSuccess
  | typeof toggleTranslation
>;
