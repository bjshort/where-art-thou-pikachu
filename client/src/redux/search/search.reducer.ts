import {
  SearchActions,
  UPDATE_QUERY,
  FETCH_POKEMON_SUCCESS,
} from './search.actions';
import { merge } from 'lodash';

export interface SearchState {
  query: string;
  result?: number;
}

const initalState: SearchState = {
  query: '',
};

export default (
  state: SearchState = initalState,
  action: SearchActions,
): SearchState => {
  switch (action.type) {
    case FETCH_POKEMON_SUCCESS: {
      return merge(state, { result: action.payload });
    }
    case UPDATE_QUERY: {
      return {
        ...state,
        query: action.payload,
      };
    }
    default:
      return state;
  }
};
