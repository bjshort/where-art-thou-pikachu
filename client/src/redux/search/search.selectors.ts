import { AppState } from '../root.reducer';

export const getQuery = (state: AppState) => state.search.query;

export const getResult = (state: AppState) => state.search.result;

export const getShowShakespeareanTranslation = (state: AppState) =>
  state.search.showShakespeareanTranslation;

export const getIsSearching = (state: AppState) => state.search.isSearching;
