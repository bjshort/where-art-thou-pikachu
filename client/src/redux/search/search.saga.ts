import { put, takeLatest, select, call } from 'redux-saga/effects';
import { getQuery } from './search.selectors';
import { AxiosResponse } from 'axios';
import { PokemonDTO } from '../../api-client/pokemon.dto';
import { ApiClient } from '../../api-client';
import {
  SearchActions,
  fetchPokemonFailed,
  fetchPokemonSuccess,
  FETCH_POKEMON_REQUESTED,
} from './search.actions';
import SCHEMA from '../schema';
import { syncEntitiesAndGetResults } from '../entities/entities.saga';
import { getPokemonByName } from '../entities/entities.selectors';

// ****************
// WORKERS
// ****************
export function* findPokemon() {
  try {
    const query: string = yield select(getQuery);
    const pokemonFromExistingState = yield select(getPokemonByName(query));

    /* 
      Becuase the state gets persisted in localStorage, lets not make real api calls if we don't have to 
    */
    if (pokemonFromExistingState) {
      yield put<SearchActions>(
        fetchPokemonSuccess(pokemonFromExistingState.id),
      );
      return;
    }

    const response: AxiosResponse<PokemonDTO> = yield call(
      ApiClient.search,
      query,
    );

    const id: number = yield syncEntitiesAndGetResults(
      response,
      SCHEMA.pokemon,
    );

    yield put<SearchActions>(fetchPokemonSuccess(id));
  } catch (err) {
    // Server returns 404 when no pokemon is found, which is not an error
    if (err.response.status === 404) {
      yield put<SearchActions>(fetchPokemonSuccess(undefined));
      return;
    }

    yield put<SearchActions>(
      fetchPokemonFailed(err.response.data.message || err.message),
    );
  }
}

// ****************
// WATCHERS
// ****************
export function* watchSearchSagas() {
  yield takeLatest(FETCH_POKEMON_REQUESTED, findPokemon);
}
