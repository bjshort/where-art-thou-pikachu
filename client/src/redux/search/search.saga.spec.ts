import { runSaga } from '@redux-saga/core';
import { findPokemon } from './search.saga';
import SCHEMA from '../schema';

describe('Search Sagas', () => {
  let dispatchedActions: any[];
  let mockState: any;
  let mockStore: any;

  beforeEach(() => {
    dispatchedActions = [];

    mockState = {
      search: {
        query: 'magikarp',
      },
      entities: {
        pokemon: {
          1234: {
            id: 1234,
            name: 'MAGikarp',
          },
        },
      },
    };

    mockStore = {
      getState: () => mockState,
      dispatch: (action: any) => dispatchedActions.push(action),
    };
  });
  describe('findPokemon', () => {
    it('Gets pokemon from store by name (case insensitive)', async () => {
      await runSaga(mockStore, findPokemon).toPromise();
      expect(dispatchedActions).toMatchSnapshot();
    });
  });
});
