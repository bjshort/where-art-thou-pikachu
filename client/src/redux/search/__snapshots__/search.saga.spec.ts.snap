// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Sagas findPokemon Gets pokemon from store by name (case insensitive) 1`] = `
Array [
  Object {
    "payload": 1234,
    "type": "SEARCH/FETCH_POKEMON_SUCCESS",
  },
]
`;
